types/src/write_set.rs

```diff
@@ -19,42 +22,161 @@ pub enum WriteOp {
     Creation(#[serde(with = "serde_bytes")] Vec<u8>),
     Modification(#[serde(with = "serde_bytes")] Vec<u8>),
     Deletion,
+    CreationWithMetadata {
+        #[serde(with = "serde_bytes")]
+        data: Vec<u8>,
+        metadata: StateValueMetadata,
+    },
+    ModificationWithMetadata {
+        #[serde(with = "serde_bytes")]
+        data: Vec<u8>,
+        metadata: StateValueMetadata,
+    },
+    DeletionWithMetadata {
+        metadata: StateValueMetadata,
+    },
 }
 
 impl WriteOp {
     #[inline]
     pub fn is_deletion(&self) -> bool {
         match self {
-            WriteOp::Deletion => true,
-            WriteOp::Modification(_) | WriteOp::Creation(_) => false,
+            WriteOp::Deletion | WriteOp::DeletionWithMetadata { .. } => true,
+            WriteOp::Modification(_)
+            | WriteOp::ModificationWithMetadata { .. }
+            | WriteOp::Creation(_)
+            | WriteOp::CreationWithMetadata { .. } => false,
         }
     }
 
     pub fn is_creation(&self) -> bool {
         match self {
-            WriteOp::Creation(_) => true,
-            WriteOp::Modification(_) | WriteOp::Deletion => false,
+            WriteOp::Creation(_) | WriteOp::CreationWithMetadata { .. } => true,
+            WriteOp::Modification(_)
+            | WriteOp::ModificationWithMetadata { .. }
+            | WriteOp::Deletion
+            | WriteOp::DeletionWithMetadata { .. } => false,
         }
     }
 
     pub fn is_modification(&self) -> bool {
         match self {
-            WriteOp::Modification(_) => true,
-            WriteOp::Creation(_) | WriteOp::Deletion => false,
+            WriteOp::Modification(_) | WriteOp::ModificationWithMetadata { .. } => true,
+            WriteOp::Creation(_)
+            | WriteOp::CreationWithMetadata { .. }
+            | WriteOp::Deletion
+            | WriteOp::DeletionWithMetadata { .. } => false,
+        }
+    }
+
+    pub fn squash(op: &mut Self, other: Self) -> Result<bool> {
+        use WriteOp::*;
+
+        // n.b. With write sets from multiple sessions being squashed together, it's possible
+        //   to see two ops carrying different metadata (or one with it the other without)
+        //   due to deleting in one session and recreating in another. The original metadata
+        //   shouldn't change due to the squash.
```

Comment:
Does this mean I cannot change the billing account of an item in a transaction?

Reply:
Right, you can't. The write set is the end result of the entire txn, which can express "a thing is modified" but not "a thing is deallocated and then re-created" in the first place.